slam-libraries/viam-cartographer/src/io/file_handler_test.cc

```diff
@@ -0,0 +1,183 @@
+    // and the right files are still in the directory
+    std::vector<std::string> listedFiles =
+        ListSortedFilesInDirectory(tmpdir.string());
+    BOOST_TEST(files.size() == listedFiles.size());
+    for (int i = 0; i < files.size(); i++) {
+        BOOST_TEST(files.at(i).compare(listedFiles.at(i)));
+    }
+    // Close the file and remove the temporary directory and its contents
+    boost::filesystem::remove_all(tmpdir);
+}
+
+BOOST_AUTO_TEST_CASE(TimedPointCloudDataFromPCDBuilder_success) {
+    // Create a mini PCD file and save it in a tmp directory
+    std::string filename = "rplidar_data_2022-01-01T01:00:00.0001Z.pcd";
+    std::string pcd =
+        "VERSION .7\n"
+        "FIELDS x y z rgb\n"
+        "SIZE 4 4 4 4\n"
+        "TYPE F F F I\n"
+        "COUNT 1 1 1 1\n"
+        "WIDTH 3\n"
+        "HEIGHT 1\n"
+        "VIEWPOINT 0 0 0 1 0 0 0\n"
+        "POINTS 3\n"
+        "DATA ascii\n"
+        "-0.001000 0.002000 0.005000 16711938\n"
+        "0.582000 0.012000 0.000000 16711938\n"
+        "0.007000 0.006000 0.001000 16711938\n";
+    // Create a unique path in the temp directory and add the PCD file
+    boost::filesystem::path tmpdir = boost::filesystem::temp_directory_path() /
+                                     boost::filesystem::unique_path();
+    bool ok = boost::filesystem::create_directory(tmpdir);
+    if (!ok) {
+        throw std::runtime_error("could not create directory: " +
+                                 tmpdir.string());
+    }
+    boost::filesystem::ofstream ofs(tmpdir / filename);
+    ofs << pcd;
+    ofs.close();
+    // Read it in and check if the data in the TimedPointCloudData is equivalent
+    // to what we had in the pcd file
+    cartographer::sensor::TimedPointCloudData timed_pcd =
+        TimedPointCloudDataFromPCDBuilder(tmpdir.string() + "/" + filename, 0);
+
+    BOOST_TEST(timed_pcd.ranges.size() == 3);
+    cartographer::sensor::TimedRangefinderPoint point_1 =
+        timed_pcd.ranges.at(0);
+    cartographer::sensor::TimedRangefinderPoint point_2 =
+        timed_pcd.ranges.at(1);
+    cartographer::sensor::TimedRangefinderPoint point_3 =
+        timed_pcd.ranges.at(2);
+
+    auto tolerance = boost::test_tools::tolerance(0.00001);
+    BOOST_TEST(point_1.position(0, 0) == -0.001, tolerance);
+    BOOST_TEST(point_1.position(1, 0) == 0.002, tolerance);
+    BOOST_TEST(point_1.position(2, 0) == 0.005, tolerance);
+
+    BOOST_TEST(point_2.position(0, 0) == 0.582, tolerance);
+    BOOST_TEST(point_2.position(1, 0) == 0.012, tolerance);
+    BOOST_TEST(point_2.position(2, 0) == 0.000, tolerance);
+
+    BOOST_TEST(point_3.position(0, 0) == 0.007, tolerance);
+    BOOST_TEST(point_3.position(1, 0) == 0.006, tolerance);
+    BOOST_TEST(point_3.position(2, 0) == 0.001, tolerance);
+    // Close the file and remove the temporary directory and its contents
+    boost::filesystem::remove_all(tmpdir);
+}
+
+BOOST_AUTO_TEST_CASE(ReadTimeFromFilename_success) {
```

Comment:
[q] Do we want to test a failed case of this? Or are there too many ways a data filename might be wrong to make it worth testing?

Reply:
Maybe a check with a filename that has an incorrect format?

Reply:
good idea, adding